BUG: DataSet assignment operator is missing

Leengit opened the pull request:
* Some compilers complain if the copy constructor is given explicitly but the assignment operator is implicitly set to default.
* Explicitly defining the assignment operator allows us to properly handle reference counters for shared resources.

Commit:
BUG: DataSet assignment operator is missing

* Some compilers complain if the copy constructor is given explicitly
  but the assignment operator is implicitly set to default.

* Explicitly defining the assignment operator allows us to properly
  handle reference counters for shared resources.

Leengit commented:
The force push to create the current commit 1e1e2677b5304606432422b276be2230f6157723 changes clang format only.

qkoziol commented:
@bmribler - Does this look OK?   Does it need a test?

bmribler commented on c++/src/H5DataSet.cpp:
```diff
@@ -76,6 +76,27 @@ DataSet::DataSet(const DataSet &original) : H5Object(), AbstractDs(), id(origina
+    return *this;
+}
+
 //--------------------------------------------------------------------------
```
Please consider using IdComponent::setId() to ensure the current dataset ID is properly closed.  setId() also handles the ref counting.  Perhaps, DataType::operator= can be used as an example.  Also, please add a test.  Thank you!

Leengit replied:
@bmribler Thank you for the feedback.  The new code has `H5DataSet::operator=` mimicking `H5DataType::operator=` as you suggest.

Leengit replied:
If someone could contribute a test that would be great.  I am a little over my head there.

bmribler replied:
Thank you, @Leengit.  I'll try to come up with a test.

Commit:
BUG: DataSet assignment operator is missing.

* Mimicking code of H5DataType::operator() as suggested by @bmribler.

derobins commented on c++/src/H5DataSet.cpp:
```diff
@@ -76,6 +76,21 @@ DataSet::DataSet(const DataSet &original) : H5Object(), AbstractDs(), id(origina
+// Function:    DataSet assignment operator
+///\brief       Assignment operator: same HDF5 object as \a original
+///\param       original - IN: DataSet instance to copy
+// Programmer   Lee Newberg - 2021
```
I'm not excited about programmer graffitti. I'm trying to remove author fields in comments when I find them as they are frequently meaningless. Developers should be using the history to determine this.

bmribler replied:
I will be adding test for this and will take care of the issues you mentioned.  Thanks!

derobins commented:
This also needs a RELEASE.txt note.

derobins reviewed (changes requested):
Definitely needs a RELEASE.txt entry.

bmribler committed:
Added test

Description:
    Added test for DataSet::operator= that Leengit added
Platform tested:
    Linux/64 (jelly)

bmribler committed:
Removed Author field.

lrknox committed:
Commit clang format changes.

lrknox commented on c++/test/dsets.cpp:
```diff
@@ -1315,6 +1288,63 @@ test_virtual()
+        DataSet copied_dataset = another_dataset;
+
+        hid_t copied_dataset_id  = copied_dataset.getId();
+        hid_t another_dataset_id = another_dataset.getId();
```
I built and ran the test, and these two lines caused these unused-variable warnings:
/home/hdftest/snapshots-hdf5_1_12_1/current/c++/test/dsets.cpp: In function ‘herr_t test_operator(H5::H5File&)’:
/home/hdftest/snapshots-hdf5_1_12_1/current/c++/test/dsets.cpp:1329:15: warning: unused variable ‘copied_dataset_id’ [-Wunused-variable]
         hid_t copied_dataset_id  = copied_dataset.getId();
               ^
/home/hdftest/snapshots-hdf5_1_12_1/current/c++/test/dsets.cpp:1330:15: warning: unused variable ‘another_dataset_id’ [-Wunused-variable]
         hid_t another_dataset_id = another_dataset.getId();

Can you use the ids for some trivial purpose?  maybe to close them, or to get the names?

bmribler committed:
Entry for Leengit's github PR #503

bmribler committed:
Merge branch 'DataSet_assignment_operator' of https://github.com/Leengit/hdf5 into DataSet_assignment_operator

bmribler committed:
Removed lines left by mistake

Leengit commented on c++/src/H5DataSet.cpp:
```diff
@@ -710,11 +698,10 @@ DataSet::getId() const
+// brief        Reads a fixed length \a H5std_string from a dataset.
+// param        mem_type  - IN: DataSet datatype (in memory)
+// param        strg      - IN: Buffer for read string
+// exceptio n   H5::DataSetIException
```
`exception`

Leengit reviewed (commented):
Thank you!
